sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala

```diff
@@ -165,4 +167,31 @@ object ExtractPythonUDFs extends Rule[SparkPlan] {
       }
     }
   }
+
+  // Split the original FilterExec to two FilterExecs. The upper FilterExec only contains
+  // Python UDF and non-deterministic predicates.
+  private def trySplitFilter(plan: SparkPlan): SparkPlan = {
+    plan match {
+      case filter: FilterExec =>
+        // Only push down the predicates that is deterministic and all the referenced attributes
+        // come from child.
+        val (candidates, containingNonDeterministic) =
+          splitConjunctivePredicates(filter.condition).span(_.deterministic)
+        val (pushDown, rest) = candidates.partition(!hasPythonUDF(_))
```

Comment:
This will change the semantics. `span` and `partition` have different semantics. Thus, we still have to keep the existing behavior. 

Let me write a comment to explain `PythonUDF` is always assumed to deterministic.